archive/compression/compression_test.go

```diff
@@ -81,3 +81,58 @@ func TestCompressDecompressGzip(t *testing.T) {
 func TestCompressDecompressUncompressed(t *testing.T) {
 	testCompressDecompress(t, 1024*1024, Uncompressed)
 }
+
+func generateCompressedData(b *testing.B, sizeInMb int) []byte {
+	size := sizeInMb * 1024 * 1024
+	b.SetBytes(int64(size))
+
+	// Make results predictable
+	rand.Seed(1234)
+
+	data := make([]byte, size)
+	if _, err := rand.Read(data); err != nil {
+		b.Fatal(err)
+	}
+
+	var buf bytes.Buffer
+	compressor, err := CompressStream(&buf, Gzip)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	if n, err := compressor.Write(data); err != nil || n != size {
+		b.Fatal(err)
+	}
+
+	compressor.Close()
+	return buf.Bytes()
+}
+
+func benchmarkDecompression(sizeInMb int) func(*testing.B) {
+	return func(b *testing.B) {
+		compressed := generateCompressedData(b, sizeInMb)
+
+		b.ResetTimer()
+		for n := 0; n < b.N; n++ {
+			decompressor, err := DecompressStream(bytes.NewReader(compressed))
+			if err != nil {
+				b.Fatal(err)
+			}
+
+			_, err = ioutil.ReadAll(decompressor)
```

Comment:
Try this `_, err = io.CopyBuffer(ioutil.Discard, decompressor, buf)` where buf is declared before the closure as `buf := make([]byte, 32*1024)`. I saw a massive improvement locally with that.